README
Msgidd - a Message ID daemon.

This is a part of the HUSKY portable fidonet software project.


I. Introduction.


A Msgidd package provides independent support for the generation of the MSGID
for any FTN software. Message ID daemon listens a TCP port 16890 and generate
new MSGID number for each request. Daemon also store last generated MSGID
number into disk file to prevent drops of MSGID sequence if daemon is aborted.

Msgidd is tested on GNU/Linux, FreeBSD and Windows.


II. Package contents.


The Msgidd package contains:
- msgidd
- msgidclient
- MsgIdClient.class
- README


III. Package components descriptions.


1. msgidd

This is a simple msgid server. It is based on the tserver code from the
book "Linux Application Development".

Msgidd waits for a connection on port 16890/TCP. Once a connection has been
established, get commands and return appropriate return message, and then
wait for another connection to the socket.

Now Msgidd recognizes only command "getMsgId": generate and return new MSGID.

To informational purposes please add following line into /etc/services:

msgid		16890/tcp	# husky MSGID daemon

After this you may get the new MSGID using command:

telnet localhost msgid
(put string "getMsgId")
(read generated MSGID)


2. msgidclient

This is a simple msgidd client. It is based on the tclient code from the
book "Linux Application Development".

Msgidclient connects to the server whose hostname or IP address is given as
an command line argument, at port 16980/TCP (msgidd port). Once connected,
write a string "getMsgId" to the socket, read the data from socket and
print this data to stdout until server closes a connection. If connection
can't be established then Msgidclient generates MSGID using internally
implemented algorithm.


3. MsgIdClient.class

This is a JAVA msgidd client.

MsgIdClient.class connects to the server whose hostname or IP address is
given as an command line argument, at port 16980/TCP (msgidd port). Once
connected, write a string "getMsgId" to the socket, read the data from
socket and print this data to stdout until server closes a connection.
Internal MSGID generator isn't implemented.


4. README

This file.


IV. Build from sources.


To compile programs the GNU C compiler package is needed. This compiler is
included into base FreeBSD distribution and into many Linux distributions.
For compile on Windows a Cygwin or MINGW32+MSYS is needed.

To compile programs please run these commands from your shell:

cd <to directory with unpacked sources>
make



$Id$


ChangeLog
msgidd 0.2
------------

----------------------------
2005/09/08 19:15:09;  author: stas_degteff;
Fix decimal and hex numbers mix. Prevent buffer overflow
----------------------------
2005/09/08 19:01:03;  author: stas_degteff;
generate msgid using current time if msgid file can't readed
----------------------------
2005/09/08 18:18:50;  author: stas_degteff;
Change version number of the MSGID daemon to 0.2.
----------------------------
2005/09/08 18:07:28;  author: stas_degteff;
Now msgidd store MSGID after each generation. Also compiler warnings is fixed.
----------------------------
2005/09/08 17:55:34;  author: stas_degteff;
Documents a MSGID daemon and clients. (Create a README.)
----------------------------

msgidd 0.1
------------

----------------------------
2000/09/13 18:28:26;  author: sms;
renamed makefile to Makefile
----------------------------
2000/09/13 18:28:08;  author: sms;
changed default target to all
----------------------------
2000/02/21 14:34:08;  author: mt;
removed msgid file
----------------------------
2000/02/19 12:13:33;  author: sms;
changed format of msgid-file to decimal
----------------------------
1999/10/27 19:42:10;  author: mt;
added creation of msgids in the client, when no server is available
----------------------------
1999/10/27 17:25:44;  author: mt;
host parameter added to java client
----------------------------
1999/10/25 11:03:43;  author: mt;
added java client
----------------------------
1999/10/25 09:05:37;  author: mt;
created makefile


msgidclient.c
/* msgidclient.c - a main file of the
   simple msgid client for unix enviroments.
   It is based on the tclient code from the book
   "Linux Application Development" */

/* Msgidclient connects to the server whose hostname or IP address is given as
   an command line argument, at port 16980/TCP (msgidd port). Once connected,
   write a string "getMsgId" to the socket, read the data from socket and print
   this data to stdout until server closes a connection. */

/* $Id$ */

#include <arpa/inet.h>
#include <netdb.h>
#include <netinet/in.h>
#include <stdio.h>
#include <sys/socket.h>
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include <stdlib.h>
#include <string.h>

#include "sockutil.h"          /* some utility functions */
#include "msgid.h"             /* createMsgId() */


int getMsgId(char *hostName, char *msgId)
{
    struct sockaddr_in address;
    struct in_addr inaddr;
    struct hostent * host;
    int sock, rc;

    memset(msgId, 0, 9);

    /* If the argument can be converted to an IP, do so. If not, try
       to look it up in DNS. */
    if (inet_aton(hostName, &inaddr))
        host = gethostbyaddr((char *) &inaddr, sizeof(inaddr),
                               AF_INET);
    else
        host = gethostbyname(hostName);

    if (!host) {
	/* We can't find an IP number */
        herror("error looking up host");
	exit(1);
    }

    if ((sock = socket(PF_INET, SOCK_STREAM, 0)) < 0)
        die("socket");

    address.sin_family = AF_INET;
    address.sin_port = htons(16890);

    /* Take the first IP address associated with this hostname */
    memcpy(&address.sin_addr, host->h_addr_list[0],
           sizeof(address.sin_addr));

    if (connect(sock, (struct sockaddr *) &address, sizeof(address)))
      {
	printf("Could not connect to msgid Server, create my own msgid\n");
	sprintf(msgId, "%08lX", createMsgId()&0xFFFFFFFF);
	rc = 1;
      }
    else {
      printf("Connected to the msgid Server, getting unique msgid\n");
      write(sock, "getMsgId", 8);
      read(sock, msgId, 8);
      close(sock);
      rc = 0;
    }

    return rc;
}

int main(int argc, char ** argv) {
  int rc;
  char msgId[9];
  int i;

  if (argc != 2) {
    fprintf(stderr, "only a single argument is supported\n");
    return 1;
  }

  for (i = 0; i < 1 ; i++)
      {
	  rc = getMsgId(argv[1], msgId);
	  printf("Got msgId: %s\n", msgId);
	  printf("rc:        %d\n", rc);
      }

  return rc;
}


msgidd.c
/* simple MSGID server for unix enviroments
   it is based on the tserver code from the book
   "Linux Application Developmend" */

/* Waits for a connection on port husky = 16890. Once a connection has been
   established, get commands and return appropriate return message, and then
   wait for another connection to the socket.
   Last generated MSGID is stored into disk file "msgid" and daemon reads it
   at start time.
 */

/* $Id$ */

#include <arpa/inet.h>
#include <netinet/in.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <sys/socket.h>
#include <unistd.h>
#include <stdlib.h>
#include <process.h>
#include <errno.h>

#include "sockutil.h"          /* some utility functions */
#include "msgid.h"             /* createMsgId() */


#define VERSION "0.2"             /* Daemon version */
#define MSGID_STORE_FILE "msgid"  /* disk file to store last generated MSGID value */

volatile int terminate = 0;
static unsigned long msgid = 0;

void termHandler(int signum)
{
  terminate = 1;
}

void writeToFile()
{
  FILE *f;

  f = fopen(MSGID_STORE_FILE, "w");
  if (f != NULL)
    {
      printf("Writing msgId File \"%s\"\n", MSGID_STORE_FILE);
      fprintf(f, "%lu", msgid);
      fclose(f);
    }
  else
    {
      printf("Can't write msgId File: %s\n", strerror(errno));
    }
}

void loadFromFile()
{
   FILE *f;
   char buf[9];

   f = fopen(MSGID_STORE_FILE, "r");
   if (f != NULL)
     {
       buf[0]=0;
       printf("Reading msgId File..\n");
       fgets(buf, 9, f);
       msgid = strtoul(buf, NULL, 10);
       fclose(f);
     }
   else
     {
       printf("Can't load msgId from file, generate new.\n");
     }
   if(msgid==0) msgid = createMsgId();
   printf("Starting with msgid: %08lX\n", msgid);
}

void installSignalHandlers()
{
  struct sigaction sa;

  memset(&sa, 0, sizeof(sa));
  sa.sa_handler = termHandler;
  if (sigaction(SIGTERM, &sa, NULL)) perror("sigaction");

  memset(&sa, 0, sizeof(sa));
  sa.sa_handler = termHandler;
  if (sigaction(SIGINT, &sa, NULL)) perror("sigaction");
}

int main(void) {
    struct sockaddr_in address;
    int sock, conn, i, rc;
    size_t addrLength = sizeof(struct sockaddr_in);
    char inBuf[12], outBuf[9];

    printf("Husky MsgId Server v.%s\n",VERSION);

    loadFromFile();

    if ((sock = socket(PF_INET, SOCK_STREAM, 0)) < 0)
        die("socket");

    /* Let the kernel reuse the socket address. This lets us run
       twice in a row, without waiting for the (ip, port) tuple
       to time out. */
    i = 1;					
    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &i, sizeof(i));

    address.sin_family = AF_INET;
    address.sin_port = htons(16890);
    memset(&address.sin_addr, 0, sizeof(address.sin_addr));

    if (bind(sock, (struct sockaddr *) &address, sizeof(address)))
        die("bind");

    if (listen(sock, 5))
        die("listen");

    installSignalHandlers();

    while ((conn = accept(sock, (struct sockaddr *) &address,
                          &addrLength)) >= 0) {

      if (terminate == 1) break;

      memset(&inBuf, 0, sizeof(inBuf));
      printf("---- getting data\n");
      rc = read(conn, &inBuf, sizeof(inBuf)-1);
      //printf("%u : %s\n", rc, inBuf);
      //for (i=0; i < rc; i++) printf("%d\n", inBuf[i]);
      if (strncmp(inBuf, "getMsgId", 8)==0)
	{
	  sprintf(outBuf, "%08lX", (msgid++)&0xffffffff);
	  printf("Generated msgid: %s\n", outBuf);
	  writeToFile();
	  write(conn, outBuf, strlen(outBuf));
	}
      else {
	printf("unknown data\n");
	write (conn, "error", 6);
      }
      printf("---- done\n");
      close(conn);
    }

    close(sock);
    return 0;
}
